src/uu/tail/src/chunks.rs

```diff
@@ -81,3 +91,630 @@ impl<'a> Iterator for ReverseChunks<'a> {
+    /// over these chunks. If there are no chunks, for example because the piped stdin contained no
+    /// bytes, or `num_print = 0` then `iterator.next` returns None.
+    ///
+    /// # Arguments
+    ///
+    /// * `reader`: A [`BufReader`] with an inner element implementing the [`Read`] trait.
+    ///
+    /// returns: Result<Iter<Box<BytesChunk, Global>>, Box<dyn UError, Global>>
+    ///
+    /// # Examples
+    ///
+    /// ```rust,ignore
+    /// use std::io::{BufReader, Cursor};
+    ///
+    /// let mut reader = BufReader::new(Cursor::new(""));
+    /// let num_print = 0;
+    /// let mut chunks = BytesChunkBuffer::new(num_print);
+    /// let mut iter = chunks.fill(&mut reader).unwrap();
+    ///
+    /// let chunk = iter.next();
+    /// assert!(chunk.is_none());
+    ///
+    /// let mut reader = BufReader::new(Cursor::new("a"));
+    /// let num_print = 1;
+    /// let mut chunks = BytesChunkBuffer::new(num_print);
+    /// let mut iter = chunks.fill(&mut reader).unwrap();
+    ///
+    /// let chunk = iter.next();
+    /// assert!(chunk.is_some());
+    /// assert_eq!(&[b'a'], chunk.unwrap().get_buffer());
+    /// assert_eq!(None, iter.next());
+    /// ```
+    pub(crate) fn fill(
+        &mut self,
+        reader: &mut BufReader<impl Read>,
+    ) -> UResult<Iter<Box<BytesChunk>>> {
+        let mut chunk = Box::new(BytesChunk::new());
+
+        // fill chunks with all bytes from reader and reuse already instantiated chunks if possible
+        while (chunk.fill(reader)?).is_some() {
+            self.bytes += chunk.bytes as u64;
+            self.chunks.push_back(chunk);
+
+            let first = &self.chunks[0];
+            if self.bytes - first.bytes as u64 > self.num_print {
+                chunk = self.chunks.pop_front().unwrap();
+                self.bytes -= chunk.bytes as u64;
+            } else {
+                chunk = Box::new(BytesChunk::new());
+            }
+        }
+
+        // quit early if there are no chunks for example in case the pipe was empty
+        if self.chunks.is_empty() {
+            return Ok(self.chunks.iter());
+        }
+
+        let chunk = self.chunks.pop_front().unwrap();
+        // calculate the offset in the first chunk and put the calculated chunk as first element in
+        // the self.chunks collection.
+        let offset = if self.num_print >= self.bytes {
+            // ignore a passed in value exceeding the number of actually read bytes and treat it
+            // like a value equal to the number of bytes.
+            0
+        } else {
+            // the calculated offset must be in the range 0 to BUFFER_SIZE and is therefore safely
+            // convertible to a usize without losses.
+            (self.bytes - self.num_print) as usize
+        };
+
+        self.chunks
+            .push_front(Box::new(BytesChunk::from_chunk(&chunk, offset)));
+
+        Ok(self.chunks.iter())
+    }
+}
+
+#[derive(Debug)]
```

Comment:
Two very nitpicky comments, but I like the doc comment above the derives and `pub(crate)` and `pub` are used for similar things in this file. Generally, because `uutils` is not a library, simply `pub` is good enough in my opinion.